magento/magento2
# ISSUE-27590 - Improve downloadable link status checks on order placement

### Description (*)

When an order containing downloadable products is placed, observers are executed to create downloadable link records (downloadable_link_purchased_item). Various criteria are checked to determine what the link statuses should be. The user will be unable to access the links until they are marked as available.

Downloadable products can be configured to become available only after the items are invoiced. However, under certain circumstances, the links are still unavailable even after being invoiced. This is because some of the checks are based on order status instead of item status. As an example, an order containing downloadable products and physical items won't reach Complete status until all items are shipped. This means the download links remain unavailable until all items are shipped.

This PR changes the initial check during download link creation to evaluate individual item status instead of order status. If the order item is invoiced, it's marked as available. This resolves issue with both mixed orders (containing downloadable and physical) and zero dollar orders (where the order status skips straight to closed instead of reaching complete).

### Fixed Issues (if relevant)
1. Fixes https://github.com/magento/magento2/issues/27590
2. Fixes https://github.com/magento/magento2/issues/25177

### Manual testing scenarios (*)
Scenario 1:
1. In store configuration > Catalog / Catalog > Downloadable Product Options, set Order Item Status to Enable Downloads = Invoiced.
1. Configure a payment method that will invoice on order placement, such as Braintree configured with Auth & Capture payment action.
1. Create a downloadable product with downloadable links.
1. Log in to the store front.
1. Add the downloadable product to the cart as well as at least one physical item.
1. Check out, pay with the payment method mentioned above, and place the order.
1. Go to /downloadable/customer/products/ and confirm the appropriate download links are available.

Scenario 2:
1. In store configuration > Sales / Payment Methods > Zero Subtotal Checkout, set New Order Status = Processing and Automatically Invoice All Items = Yes.
1. Perform steps 1 - 4 from scenario 1, but price the downloadable product as $0.
1. On the store front, add the downloadable product to the cart (with no other items), check out, and place the order.
1. Go to /downloadable/customer/products/ and confirm the appropriate download links are available.


### Questions or comments


### Contribution checklist (*)
